### PR TITLE
Triplet Loss Solution to 2020 Kaggle ISIC Challenge (47034073)

### PR DESCRIPTION
**Pull request by MitchellJC**
Author: Mitchell Clark -- 47034073

This pull request features a solution to the 2020 Kaggle ISIC Challenge. The main files are

- `dataset.py`: Contains PyTorch datasets
- `modules.py`: Contains Pytorch modules. Mainly a slightly modified ResNet for creating embeddings.
- `trainer.py`: Contains trainer class for running training loops, saving and loading the model and training data.
- `train.py`: Contains code for initializing training and classifiers. Contains plotting code.
- `predict.py`: Can be used to predict if a single, user-chosen, image of a lesion is benign or malignant.

Other Python files are data utilities and general utilities which should not need to be used if the provided dataset in the README is downloaded.

See the `README.md` for a more in-depth discussion of the solution.

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
Model is restnet with an embedding layer. Does a classification only (no detection)

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **18**

Solves problem: The solution is well-suited to the problem (**3**).
The training manifold visualization,  confusion matrix, losses,ROC , and accuracy curves are provided.
Triplet loss is utilized with a hard mining strategy, which is suitable for the problem.
Class imbalance is addressed through undersampling of benign cases.

The SVM achieves 90% accuracy with 91% specificity and 44% sensitivity, while the KNN achieves 86% accuracy with 87% specificity and 54% sensitivity

Implementation functions as intended (**3**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

**Notes:**
-The AUC progression from 0.61 (training) to 0.81 (validation) and 0.84 (test) suggests that the model is improving across splits, but the low training AUC (0.61) may indicate underfitting on the training data.
_possibly improve the model architecture, tuning hyperparameters, or training for more epochs and regularize carefully to avoid overfitting while increasing training performance._
- Reasonable class separability in the feature space, as seen in the tSNE plot, though there is some overlap.
- Despite using a balanced dataset, the models exhibit unbalanced sensitivity and specificity, with high specificity but lower sensitivity, indicating that they still struggle to detect malignant cases effectively
- Using an algorithmic approach to undersample the majority class is indeed recommended over random sampling, especially in the context of medical imaging to maintain informative samples, improve generalization, and avoide loss of critical information.
- The performance of the AUC and loss is not entirely consistent. The low training AUC alongside a potentially stable loss curve suggests that the model could be improving loss without effectively learning distinguishing features in the training set. This inconsistency could be addressed by revisiting the model architecture, training strategy, or hyperparameters.

**Comment by MitchellJC** [comment]
> ### This is an initial inspection, no action is required at this point
> Recognition Problem : total : **18**
> 
> Solves problem: The solution is well-suited to the problem (**3**). The training manifold visualization, confusion matrix, losses,ROC , and accuracy curves are provided. Triplet loss is utilized with a hard mining strategy, which is suitable for the problem. Class imbalance is addressed through undersampling of benign cases.
> 
> The SVM achieves 90% accuracy with 91% specificity and 44% sensitivity, while the KNN achieves 86% accuracy with 87% specificity and 54% sensitivity
> 
> Implementation functions as intended (**3**)
> 
> Good design: Well-designed (**1**)
> 
> Commenting: Clear and sufficient comments throughout the code. (**1**)
> 
> Difficulty: Hard (**10**)
> 
> **Notes:** -The AUC progression from 0.61 (training) to 0.81 (validation) and 0.84 (test) suggests that the model is improving across splits, but the low training AUC (0.61) may indicate underfitting on the training data. _possibly improve the model architecture, tuning hyperparameters, or training for more epochs and regularize carefully to avoid overfitting while increasing training performance._
> 
> * Reasonable class separability in the feature space, as seen in the tSNE plot, though there is some overlap.
> * Despite using a balanced dataset, the models exhibit unbalanced sensitivity and specificity, with high specificity but lower sensitivity, indicating that they still struggle to detect malignant cases effectively
> * Using an algorithmic approach to undersample the majority class is indeed recommended over random sampling, especially in the context of medical imaging to maintain informative samples, improve generalization, and avoide loss of critical information.
> * The performance of the AUC and loss is not entirely consistent. The low training AUC alongside a potentially stable loss curve suggests that the model could be improving loss without effectively learning distinguishing features in the training set. This inconsistency could be addressed by revisiting the model architecture, training strategy, or hyperparameters.

Hi Hanem,

What did I lose the two marks for in the first criterion? 

In response to the notes about inconsistent training AUC, I do mention this in the report, but the inconsistency comes from rebalancing the data to avoid overfitting, the training set for the classifier does not have the ground truth distribution whereas the training set for the embedder and validation and test set does. I mentioned that I tried resampling for the embedder but that it did not actually work well and instead only rebalanced for the classifier. I achieved an AUC of 1 alongside perfect scores for recall/precision on the train set when I did not rebalance the training set before applying the classifier so I don't think this is truly underfitting in the way mentioned. I agree that there was unbalanced sensitivity and specificity but this was not mentioned as the goal of the task so I did not try to optimize for this, I was able to achieve more balanced outputs of these at the cost of AUC.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Despite using a balanced dataset, the models exhibit unbalanced sensitivity and specificity.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splits. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. Somewhat minimal.</td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Correction on marking, missed -1 mark above

**Comment by MitchellJC** [comment]
> Correction on marking, missed -1 mark above

Hi Shakes,

What did I lose the additional mark for?

**Comment by shakes76** [comment]
Reverting deduction for model design +1, as being too harsh. The other -1 deduction is from "models exhibit unbalanced sensitivity and specificity.". This means the total deduction is -2 as originally posted by Gayan.
